### PR TITLE
kni and mempool stats

### PR DESCRIPTION
**Pull request by drunkirishcoder**
sample output of KNI stats

```
---
nb2:
  kni:
    "dropped{kni=\"kni0\",dir=\"tx\"}": 0
    "octets{kni=\"kni0\",dir=\"rx\"}": 6460
    "octets{kni=\"kni0\",dir=\"tx\"}": 0
    "packets{kni=\"kni0\",dir=\"rx\"}": 64
    "packets{kni=\"kni0\",dir=\"tx\"}": 0
  port:
    "dropped{port=\"kni0\",dir=\"rx\"}": 0
    "dropped{port=\"kni0\",dir=\"tx\"}": 0
    "errors{port=\"kni0\",dir=\"rx\"}": 0
    "errors{port=\"kni0\",dir=\"tx\"}": 0
    "no_mbuf{port=\"kni0\",dir=\"rx\"}": 0
    "octets{port=\"kni0\",dir=\"rx\",core=\"1\"}": 0
    "octets{port=\"kni0\",dir=\"tx\",core=\"1\"}": 6460
    "packets{port=\"kni0\",dir=\"rx\",core=\"1\"}": 0
    "packets{port=\"kni0\",dir=\"tx\",core=\"1\"}": 64
```

sample output of mempool stats

```
---
nb2:
  mempool:
    "free{pool=\"mempool0\"}": 65151
    "used{pool=\"mempool0\"}": 384
```

### CODE REVIEW

**Comment by tobz** [comment]
Hey y'all!  Toby here from the @metrics-rs team.

Just wanted to throw it out there that we're on Discord -- https://discord.gg/uVS97z -- if you have any deep thoughts, concerns, requests, etc, for the `metrics` ecosystem.. but issues work too. 😄 

Hope it's working well for you so far. 👍

File: core/src/dpdk/mempool.rs [hunk 1]
```diff
@@ -68,6 +67,11 @@ impl Mempool {
     pub fn name(&self) -> &str {
         self.raw().name[..].as_str()
     }
+
+    #[cfg(feature = "metrics")]
+    pub fn stats(&self) -> super::MempoolStats {
```
Comment (zeeshanlakhani) [inline core/src/dpdk/mempool.rs]: minor: why not just import in MempoolStats here?

File: core/src/metrics.rs [hunk 2]
```diff
@@ -63,6 +84,14 @@ pub(crate) fn register_port_stats(ports: &[Port]) {
     });
 }
 
+/// Registers collected mempool stats with the metrics store.
+pub(crate) fn register_mempool_stats<'a>(mempools: impl Iterator<Item = &'a Mempool>) {
+    let stats = mempools.map(Mempool::stats).collect::<Vec<_>>();
+    SINK.clone().proxy("mempool", move || {
```
Comment (zeeshanlakhani) [inline core/src/metrics.rs]: reading the docs, SINK's can be cheaply cloned yeah? I should have brought this up in a previous PR.

File: examples/kni/main.rs [hunk 3]
```diff
@@ -1,11 +1,21 @@
...
 use nb2::settings::load_config;
 use nb2::{batch, Result, Runtime};
+use std::time::Duration;
 use tracing::{debug, Level};
 use tracing_subscriber::fmt;
 
+fn print_stats() {
+    let mut observer = YamlBuilder::new().build();
+    metrics::global().controller().observe(&mut observer);
+    println!("{}", observer.drain());
```
Comment (zeeshanlakhani) [inline examples/kni/main.rs]: Also related to the syn flood example, maybe it's the Yamlbuilder? but we can print things w/o the escape symbols? It's kinda lame.

File: core/src/dpdk/stats.rs [hunk 4]
```diff
@@ -117,3 +118,55 @@ impl PortStats {
...
+                    "pool" => self.name().to_string(),
+                ),
+            ),
+            Measurement::Gauge(value),
+        )
+    }
+
+    /// Collects the mempool stats.
+    pub fn collect(&self) -> Vec<(Key, Measurement)> {
+        let used = unsafe { ffi::rte_mempool_in_use_count(self.raw()) as i64 };
```
Comment (zeeshanlakhani) [inline core/src/dpdk/stats.rs]: with the unsafe call in here, should we do anything around wrapping this as a result w/ a possible error?

File: core/src/dpdk/mempool.rs (hunk 1 above)
Comment (drunkirishcoder) [reply core/src/dpdk/mempool.rs]: cuz then you have to feature gate the import. relatively minor but spreading all these `cfg` all over is a major pita. for one import I figured why not just skip the extra import. I can certainly go back to that if that's preferred.

Comment (zeeshanlakhani) [reply core/src/dpdk/mempool.rs]: I figured, but was just thinking consistency. Not a biggie for me in terms of preference though.

File: core/src/dpdk/stats.rs (hunk 4 above)
Comment (drunkirishcoder) [reply core/src/dpdk/stats.rs]: doc didn't say this call can fail. should be safe.

File: core/src/metrics.rs (hunk 2 above)
Comment (drunkirishcoder) [reply core/src/metrics.rs]: yeah. relatively cheap as long as we are not doing it in a critical path like per packet. this should be fine.

File: examples/kni/main.rs (hunk 3 above)
Comment (drunkirishcoder) [reply examples/kni/main.rs]: unfortunately it's quote inside a quoted string, can't avoid it. we can definitely make a different observer that's more console print friendly. the current built in ones are kinda messy each in their own way.

Comment (zeeshanlakhani) [reply examples/kni/main.rs]: yeah, i think i rather output json anyhow, tbh, as the default, but hopefully we can clean this up eventually.

Comment (drunkirishcoder) [reply examples/kni/main.rs]: I tried json already, it's not pretty formatted, no line breaks. way harder to read. prometheus format has too many line breaks hard to tell the iterations apart. yaml unfortunately is the most readable one out of the built in 3 until we do a console observer ourselves.

Comment (zeeshanlakhani) [reply examples/kni/main.rs]: huh. ok i guess.

**Pull request merged**